Change device selection to MultiSelect

georgehager opened the pull request:
# FRONTEND PULL REQUEST

## Related Issue #4269 #4098 #4095

- Use a MultiSelect for device selection on the Manage Facilities page

## Changes Proposed

- [x] Replaces the device selection table (multiple dropdowns + trashcans) with a single multiselect
- [x] Fixes up error handling for device selection
  - Removed the existing error handling on `ManageDevices` component - this implementation was not tied to the rest of the form errors on the page, and was not properly accessible. Error handling now works like the rest of the form and focuses properly.
  - functionally, nothing changed on the multiselect component wrt error handling, just added a css class to render the error more prettily
  - device validation happens _only_ on `Save changes` for the facility page, see notes below
  - changing device selection clears existing errors
 - [x] Fixes up accessibility issues
   - #4098 MultiSelect replaces the former trash can icons with `X` on pills. Modified the pills to use buttons and proper aria elements for accessibility.
   - #4095 Refactored the MultiSelect to use a valid aria role (combobox) - selection now works via keyboard - and aria label
   - Refactored the pill container to use a fieldset and an sr-only legend so the grouping now makes sense to screenreaders
 - [x] Frontend tests
   - snapshot tests - updates snaps for both facility/device management and for admin pages, since the multiselect component changed and is also featured on admin pages 
   - tests for multiselect, manage devices, and manage facility

## Additional Information

- Error validation - it was not working correctly before, this seemed like the right time to fix it. Form error caused by selecting no devices did not clear after device selection, it now does. Note that validation is only on `Save changes` for the facility page. This is because the fancy facility form validation does not support multiselect dropdown item arrays, and adding this custom capability would have made the code even more obtuse.  I tried adding it to the custom `onBlur` for MultiSelect like on other fields on the page, however this got very messy. I additionally did not want to break the boundary between `ManageDevices` and `MultiSelect`, and there was not a clean way to push the validation down.
- Accessibility should now be better for any pages that use `MultiSelect`! 🎉 

## Testing

- Check out the `Manage Devices` page in `dev.simplereport.gov` and ensure that the MultiSelect component is fully accessible
- I ran this implementation through AXE DevTools and there are no longer any flagged issues, but it would be great to have this confirmed

## Screenshots / Demos

- Mouse navigation of device addition/deletion (+ error handling)

https://user-images.githubusercontent.com/89797785/193041370-0330f9f8-f00e-4775-8d37-078ce85e835c.mov

- Accessible navigation of device addition/deletion (+ error handling)

https://user-images.githubusercontent.com/89797785/193041278-cf912d74-d42a-45ca-b84e-7c86846b660e.mov

- Accessibility concerns
  - has error id, see #4329 
![Screen Shot 2022-09-29 at 9 30 15 AM](https://user-images.githubusercontent.com/89797785/193045199-133e326f-1169-4539-ac98-09a97ef886aa.png)

- delete icon is visible to screenreaders, see #4098
![Screen Shot 2022-09-29 at 9 34 47 AM](https://user-images.githubusercontent.com/89797785/193046012-18a7cdd7-16de-4a12-bab7-66010cacd2dc.png)
![Screen Shot 2022-09-29 at 10 47 45 AM](https://user-images.githubusercontent.com/89797785/193063969-cd013b00-8f3a-4a0e-8b73-8926a9ddaf07.png)

- MultiSelect has a valid aria-role and aria-label, see #4095, and selections are in an accessible fieldset
![Screen Shot 2022-09-29 at 10 29 47 AM](https://user-images.githubusercontent.com/89797785/193059498-31466d15-d963-4282-a328-212ce507095a.png)

![Screen Shot 2022-09-29 at 10 47 45 AM](https://user-images.githubusercontent.com/89797785/193063969-cd013b00-8f3a-4a0e-8b73-8926a9ddaf07.png)


## Checklist for Author and Reviewer
### Design
- [ ] Any UI/UX changes have a designer as a reviewer, and changes have been approved - @kenieh 
- [ ] Any large-scale changes have been deployed to `test`, `dev`, or `pentest` and smoke-tested by both the engineering and design teams
- [ ] engineers - please review the Chromatic changes!

### Accessibility
- [ ] Any changes do not cause accessibility regressions


<!---


### Content
- [ ] Any content changes have been approved by content team

### Support
- [ ] Any changes that might generate new support requests have been flagged to the support team
- [ ] Any changes to support infrastructure have been demo'd to support team

### Security
- [ ] Changes with security implications have been approved by a security engineer (changes to  authentication, encryption, handling of PII, etc.)
- [ ] Any dependencies introduced have been vetted and discussed

### Documentation
- [ ] Any changes to the startup configuration have been documented in the README
-->

emmastephenson commented:
Overall this looks SO much better, thank you George! 🎉 I'm psyched to move away from our wonky old device selection haha.

I did notice a small regression when testing on `dev` - devices that have already been selected still appear in the select dropdown. Here's an example: 
https://watch.screencastify.com/v/1UD04mmpH7cdfuMQUWwb

I think this is a change in the behavior of MultiSelect, because looking at the superadmin edit device page it's also happening for swab and disease selection.

emyl3 reviewed (approved):
Agreed with emma! This looks awesome!!! Thank you for all your thorough testing and work on this, George! 😹 

If I understood your bug correctly, I think I see this behavior on `main` as well. I've noticed another multiselect issue that's also present in `main`. It has to do with removing an option that was previously saved before and then seeing double of that option in the select dropdown. I think we can create a separate story to address these preexisting bugs.

https://user-images.githubusercontent.com/20211771/193183271-36919d15-a689-4124-a030-9dc2e3aa565f.mov

georgehager commented:
> I did notice a small regression when testing on `dev` - devices that have already been selected still appear in the select dropdown. Here's an example: https://watch.screencastify.com/v/1UD04mmpH7cdfuMQUWwb
> 
> I think this is a change in the behavior of MultiSelect, because looking at the superadmin edit device page it's also happening for swab and disease selection.

Let me take a look at this for device selection. On my branch, I am seeing the filtering work fine for the super admin multiselects, it may be that I just need to implement a filtering function for this use case.

emmastephenson commented:
@georgehager @emyl3 the initial page load piece is very sneaky! Agreed, we can make that into a separate ticket, no need to address here.

It is still odd that filtering isn't working for device selection 🤔 The filtering function looks like it's part of MultiSelect, but none of your code touches that: https://github.com/CDCgov/prime-simplereport/blob/1f0c4e24b8c16a43926e1450ae78d530970d11e1/frontend/src/app/commonComponents/MultiSelect/MultiSelect.tsx#L87-L87

BobanL commented:
> @georgehager @emyl3 the initial page load piece is very sneaky! Agreed, we can make that into a separate ticket, no need to address here.
> 
> It is still odd that filtering isn't working for device selection 🤔 The filtering function looks like it's part of MultiSelect, but none of your code touches that:
> 
> https://github.com/CDCgov/prime-simplereport/blob/1f0c4e24b8c16a43926e1450ae78d530970d11e1/frontend/src/app/commonComponents/MultiSelect/MultiSelect.tsx#L87-L87

It seems like the use effect is setting options as available options rather than using avialbleOptions.  But I'm not toally even sure what this effect is supposed to do (maybe force sort before showing it?)
https://github.com/CDCgov/prime-simplereport/blob/1f0c4e24b8c16a43926e1450ae78d530970d11e1/frontend/src/app/commonComponents/MultiSelect/MultiSelect.tsx#L99-L102



and then for the issue that the list is containing already selected values, it just needs to be filtered when setting the initial available options 
https://github.com/CDCgov/prime-simplereport/blob/1f0c4e24b8c16a43926e1450ae78d530970d11e1/frontend/src/app/commonComponents/MultiSelect/MultiSelect.tsx#L73-L75

BobanL reviewed (approved):
Nice job on this!

emmastephenson commented on frontend/src/app/commonComponents/MultiSelect/MultiSelect.tsx:
```diff
@@ -39,18 +41,24 @@ type PillProps = {
+        aria-hidden={false}
+        fontSize={24}
+        icon={"times"}
+        aria-label={"Delete device"}
```
I think this aria-label should probably be a prop on the multiSelect, since the icon's purpose won't always be device deletion. "Delete `label`" maybe?

georgehager replied:
ooh thank you! good catch

georgehager replied:
fixed in the latest commit!

emmastephenson commented:
Also note, I ran the new and improved device selector through all the Deque manual tests and it passed all of them! 🎉

georgehager commented:
Ok I fixed ALL THE THINGS, this should be ready for re-review @BobanL @emyl3 and @emmastephenson.

New changes:
- removed the mysterious `useEffect` that did not seem to have a (positive) effect 
- added lodash `_uniquBy` to make sure there are no duplicated items in the dropdown list
- added some filtering for options on initial page load
- added and rearranged tests for `MultiSelect.tsx`

All the bugs discussed above should be fixed, please check on `dev1` which is just now deploying!

emmastephenson commented on frontend/src/app/commonComponents/MultiSelect/MultiSelect.test.tsx:
```diff
@@ -17,85 +17,130 @@ const fruitOptions: MultiSelectDropdownOption[] = [
+          expect(optionList.children.length).toEqual(5);
+          expect(
+            within(optionList).queryByText("Oranges")
+          ).not.toBeInTheDocument();
```
nit: I think this test is asserting the opposite of what we expect?

georgehager replied:
@emmastephenson I took a look back through the whole test suite, and this test case is correct. It's a bit hard to follow because the suite was written with multiple nested `beforeEach` statements, which sets things up such that for this last test case `Oranges` actually isn't in the list. Personally I'd like to see us [move away from so much nesting in testing](https://kentcdodds.com/blog/avoid-nesting-when-youre-testing) b/c it makes the tests really hard to read/refactor and too inder-dependent, giving us small improvements in efficiency and a big loss of readability. For the sake of not sending this PR through another review though I'm gonna leave this as is. 🙃

emmastephenson reviewed (approved):
LGTM!! 

I'll keep mulling over the mysterious `useEffect` and let you know if I have any Eureka moments. But overall this looks great, so excited to put this in front of users!

emyl3 reviewed (approved):
LGTM! Thank you George! Was able to confirm that I'm no longer seeing the previous behavior!! 😸

BobanL reviewed (approved):
looks good!
